src/poetry/inspection/lazy_wheel.py

```diff
@@ -0,0 +1,762 @@
+"""Lazy ZIP over HTTP"""
+
+from __future__ import annotations
+
+import abc
+import io
+import logging
+import re
+
+from bisect import bisect_left
+from bisect import bisect_right
+from contextlib import contextmanager
+from tempfile import NamedTemporaryFile
+from typing import TYPE_CHECKING
+from typing import Any
+from typing import BinaryIO
+from typing import ClassVar
+from typing import cast
+from urllib.parse import urlparse
+from zipfile import BadZipFile
+from zipfile import ZipFile
+
+from pkginfo import Distribution
+from requests.models import CONTENT_CHUNK_SIZE
+from requests.models import HTTPError
+from requests.models import Response
+from requests.status_codes import codes
+
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+    from collections.abc import Iterator
+
+    from requests import Session
+
+    from poetry.utils.authenticator import Authenticator
+
+
+logger = logging.getLogger(__name__)
+
+
+class HTTPRangeRequestUnsupported(Exception):
+    """Raised when the remote server appears unable to support byte ranges."""
+
+
+class UnsupportedWheel(Exception):
+    """Unsupported wheel."""
+
+
+class InvalidWheel(Exception):
+    """Invalid (e.g. corrupt) wheel."""
+
+    def __init__(self, location: str, name: str):
+        self.location = location
+        self.name = name
+
+    def __str__(self) -> str:
+        return f"Wheel '{self.name}' located at {self.location} is invalid."
+
+
+class MemoryWheel(Distribution):
```

Comment:
Could you use [`packaging.metadata`](https://packaging.pypa.io/en/stable/metadata.html) instead of `pkginfo`? I plan on dropping `pkginfo` in favor of `packaging` quite soon.

Reply:
tldr: I don't think so.

Having just tried it out, I like that it simplifies the interface because we can get rid of `MemoryWheel`.

However, I firmly believe that **we should not partially switch from pkginfo to packaging.metadata**.

`pkginfo` just keeps the string values. This results in `version` being a string and `requires_dist` being a list of strings and so on. `packaging.metadata` does some additional parsing so `version` is a `Version` and `requires_dist` is a list of `Requirement` objects, `requires_python` is a `SpecifierSet`.

And here is the issue:
A `requries_python` like `">=2.7, !=3.0.*, !=3.1.*, !=3.2.*"` is passed as is by `pkginfo` but converted to `"!=3.0.*,!=3.1.*,!=3.2.*,>=2.7"` by `packaging.metadata`. (It's equivalent but not equal.)

If I only use `packaging.metadata` here, you will get different (though equivalent) lock files depending on whether you have activated `lazy-wheel` or not. Even worse: it will not depend on the current setting, but on the setting when the data was cached. So even `poetry lock --no-update` may result in different files on different machines if the setting was different when the metadata had been cached.

Thinking about this, **we probably should even increment the cache version, when switching from `pkginfo` to `packaging.metadata`**.

If you are interested, I put a commit switching to `packaging.metadata` on a separate branch: https://github.com/radoering/poetry/commit/f0bcee4809bf7b019f68c1399b9c81178ded1247

Reply:
Rereading the docs, a solution might be to use `packaging.metadata.parse_email()` and use the resulting `RawMetadata` dict instead of the parsed `Metadata` from `packaging.metadata.from_email()`. That way, the transition might be transparent. I'll try this tomorrow.

Reply:
`RawMetadata` seems to be the right way for us. The PR is void of `pkginfo` now.